src/librustdoc/html/render/print_item.rs

```diff
@@ -1415,3 +1423,42 @@ fn document_non_exhaustive(w: &mut Buffer, item: &clean::Item) {
         w.write_str("</div>");
     }
 }
+
+fn document_type_layout(w: &mut Buffer, cx: &Context<'_>, ty_def_id: DefId) {
+    if !cx.shared.show_type_layout {
+        return;
+    }
+
+    let param_env = cx.tcx().param_env(ty_def_id);
+    let ty = cx.tcx().type_of(ty_def_id);
+    match cx.tcx().layout_of(param_env.and(ty)) {
+        Ok(ty_layout) => {
+            writeln!(w, "<h2 class=\"small-section-header\">Layout</h2>");
+            writeln!(w, "<div class=\"docblock\">");
+            writeln!(
+                w,
+                "<div class=\"warning\"><p><strong>Note:</strong> Most layout information is \
+                 completely unstable and may be different between compiler versions. \
+                 The only exception is types with certain <code>repr(...)</code> attributes. \
```

Comment:
Can we avoid showing this warning when the user passes `repr(C)` or `repr(transparent)` when the underlying type is `repr(C)`? Happy for that to go in a follow-up if it's tricky.

Reply:
Yeah, I'd rather not make this logic too complex right now. We also don't know if the library author intended for the `repr(C)` to be stable, or if they did it for some other, internal reason.

Reply:
Also, for some reason `layout_of` doesn't seem to return the `repr`; we would probably have to ask some other part of the compiler.

Reply:
> We also don't know if the library author intended for the repr(C) to be guaranteed, or if they did it for some other, internal reason.

If you remove `#[repr(C)]`, that's a breaking change, because it means you can no longer pass types through FFI. So I do think we should show the type size for repr(C). But yeah a follow-up for that sounds fine.

Reply:
```suggestion
                "<div class=\"warning\"><p><strong>Note:</strong> Most layout information is \
                 completely unstable and may be different between compiler versions. \
                 The only exception is types with certain <code>repr(C)</code>  and <code>repr(transparent)</code> attributes. \
```
Putting this now would be better, having different messages for different case may be confusing.

Reply:
I don't want rustdoc to describe the specific attributes since it may become out of date over time. We already link to the Reference's section on type layout.